pkg/controllers/networkinfo/networkinfo_controller.go

```diff
@@ -297,45 +268,114 @@ func (r *NetworkInfoReconciler) Start(mgr ctrl.Manager) error {
 		return
 	}
-	nsSet := sets.NewString()
-	for _, ns := range namespaces.Items {
-		nsSet.Insert(ns.Name)
-	}
 
-	for i := len(nsxVPCList) - 1; i >= 0; i-- {
-		nsxVPCNamespace := getNamespaceFromNSXVPC(&nsxVPCList[i])
-		if nsSet.Has(nsxVPCNamespace) {
+	for i, nsxVPC := range nsxVPCList {
+		nsxVPCNamespaceName := filterTagFromNSXVPC(&nsxVPCList[i], commonservice.TagScopeNamespace)
+		nsxVPCNamespaceID := filterTagFromNSXVPC(&nsxVPCList[i], commonservice.TagScopeNamespaceUID)
+		if nsSet.Has(nsxVPCNamespaceName) && idSet.Has(nsxVPCNamespaceID) {
 			continue
 		}
-		elem := nsxVPCList[i]
-		log.Info("GC collected nsx VPC object", "ID", elem.Id, "Namespace", nsxVPCNamespace)
+		log.Info("Garbage collecting NSX VPC object", "VPC", nsxVPC.Id, "Namespace", nsxVPCNamespaceName)
 		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, common.MetricResTypeNetworkInfo)
-		err = r.Service.DeleteVPC(*elem.Path)
-		if err != nil {
+
+		if err = r.Service.DeleteVPC(*nsxVPC.Path); err != nil {
+			log.Error(err, "Failed to delete NSX VPC", "VPC", nsxVPC.Id, "Namespace", nsxVPCNamespaceName)
 			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, common.MetricResTypeNetworkInfo)
-		} else {
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, common.MetricResTypeNetworkInfo)
+			continue
 		}
+
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, common.MetricResTypeNetworkInfo)
+		log.Info("Successfully deleted NSX VPC", "VPC", nsxVPC.Id)
+	}
+}
+
+func (r *NetworkInfoReconciler) deleteStaleVPCs(ctx context.Context, ns, id string) error {
+	isShared, err := r.Service.IsSharedVPCNamespaceByNS(ns)
+	if err != nil {
+		return fmt.Errorf("failed to check if Namespace is shared for NS %s: %w", ns, err)
+	}
+	if isShared {
+		log.Info("Shared Namespace, skipping deletion of NSX VPC", "Namespace", ns)
+		return nil
 	}
+	nsSet, idSet, err := r.listNamespaceCRsNameIDSet(ctx)
+	if err != nil {
+		log.Error(err, "Failed to list Kubernetes Namespaces")
+		return fmt.Errorf("failed to list Kubernetes Namespaces while deleting VPCs: %v", err)
+	}
+	// Retrieve stale VPCs associated with the Namespace
+	staleVPCs := r.Service.GetVPCsByNamespace(ns)
+	if len(staleVPCs) == 0 {
+		log.Info("There is no VPCs found in store, skipping deletion of NSX VPC", "Namespace", ns)
+		return nil
+	}
+	var deleteErrs []error
+	for _, nsxVPC := range staleVPCs {
+		namespaceIDofVPC := filterTagFromNSXVPC(nsxVPC, commonservice.TagScopeNamespaceUID)
+		namespaceNameofVPC := filterTagFromNSXVPC(nsxVPC, commonservice.TagScopeNamespace)
+		// if delete with Name, the id is "", `id != namespaceIDofVPC` will always be true
+		// if delete with Name and id, namespaceNameofVPC and namespaceIDofVPC will in K8s, if id == namespaceIDofVPC, we should delete it, otherwise skip deleting it
+		if nsSet.Has(namespaceNameofVPC) && idSet.Has(namespaceIDofVPC) && id != namespaceIDofVPC {
```

Comment:
It's ok to combine the cases of deleting vpc by name and by id. Just think about if it will be better if we put delete by id in another case as we do in subnet/subnetset. No strong preference.

Reply:
Updated.